### PR TITLE
LVC-221 (#2)

### PR DESCRIPTION
**Pull request by marko-yaciv**
dev
## JIRA

* [Main JIRA ticket](https://jira.softserve.academy/secure/RapidBoard.jspa?rapidView=id)

## Code reviewers

- [ ] @github_username

### Second Level Review

- [ ] @github_username

## Summary of issue

ToDo

## Summary of change

ToDo

## Testing approach

ToDo

## CHECK LIST
- [ ]  СI passed
- [ ]  Сode coverage >=95%
- [ ]  PR is reviewed manually again (to make sure you have 100% ready code)
- [ ]  All reviewers agreed to merge the PR
- [ ]  I've checked new feature as logged in and logged out user if needed
- [ ]  PR meets all conventions

### CODE REVIEW

File: RealTimeFaceMaskDetector/SQLServerDLL/SQLServer.cpp [hunk 1]
```diff
@@ -54,7 +54,7 @@ void SQLServer::Connect()
 		throw sql_error;
 	}
 }
-
+#include <iostream>
```
Comment (marko-yaciv) [inline RealTimeFaceMaskDetector/SQLServerDLL/SQLServer.cpp]: this is already removed)))

File: RealTimeFaceMaskDetector/SQLServerDLL/SQLServer.cpp [hunk 2]
```diff
@@ -75,7 +75,7 @@ void SQLServer::InsertPhoto(const Photo& photo)
 		insert_statement += "VALUES\n(\n(SELECT BulkColumn FROM OPENROWSET(BULK \'" + full_path + "\', SINGLE_BLOB) AS IMAGE),\n";
 		insert_statement += "\'" + photo.name + "\',\n";
 		insert_statement += "\'" + photo.extension + "\'\n);";
-
+		std::cout << insert_statement <<std::endl;
```
Comment (marko-yaciv) [inline RealTimeFaceMaskDetector/SQLServerDLL/SQLServer.cpp]: This is removed too

File: RealTimeFaceMaskDetector/TCPServer/WindowsService.cpp [hunk 3]
```diff
@@ -122,16 +122,18 @@ bool Service::CreateServer()
...
 	bool is_socked_created = s_socket_server.CreateListeningSocket();
-	bool is_listening_started = s_socket_server.StartListening();
-	bool is_connection_accepted = s_socket_server.AcceptConnection();
-	bool is_message_received = s_socket_server.ReceiveMessage();
+	bool is_listening_started;
+
+	std::thread listen_multiple_clients([&]() {s_socket_server.StartListening(is_listening_started); });
+	if (listen_multiple_clients.joinable())
+	{
+		listen_multiple_clients.join();
```
Comment (marko-yaciv) [inline RealTimeFaceMaskDetector/TCPServer/WindowsService.cpp:126]: Here if we call detach(), the server will be able to listen to multiple clients.
Calling join(), the server will wait until the client finishes it's work

File: RealTimeFaceMaskDetector/TCPServer/TCPServer.vcxproj [hunk 4]
```diff
@@ -124,7 +124,7 @@
       <PreprocessorDefinitions>_DEBUG;_CONSOLE;%(PreprocessorDefinitions)</PreprocessorDefinitions>
       <ConformanceMode>true</ConformanceMode>
       <LanguageStandard>stdcpp17</LanguageStandard>
-      <AdditionalIncludeDirectories>$(SolutionDir)..\3rdPartyLibs\SQLAPI\include;$(SolutionDir)SQLServerDLL;$(SolutionDir)EncryptDecryptECBMode;%(AdditionalIncludeDirectories)</AdditionalIncludeDirectories>
+      <AdditionalIncludeDirectories>$(SolutionDir)..\3rdPartyLibs\SQLAPI\include;$(SolutionDir)SQLServerDLL;$(SolutionDir)EncryptDecryptECBMode;./;%(AdditionalIncludeDirectories)</AdditionalIncludeDirectories>
```
Comment (ShostakOleh) [inline RealTimeFaceMaskDetector/TCPServer/TCPServer.vcxproj:127]: What "./" is used for?

**Review by ShostakOleh — APPROVED** [review]

File: RealTimeFaceMaskDetector/TCPServer/TCPServer.vcxproj (hunk 4 above)
Comment (marko-yaciv) [reply RealTimeFaceMaskDetector/TCPServer/TCPServer.vcxproj:127]: I had an isue that other files in that project didn't want to be included. So i added project dir to include directories

File: 3rdPartyLibs/easylogging++/samples/VC++/VCPP2015_Win32/VCPP2015_Win32/VCPP2015_Win32.vcxproj.filters [hunk 5]
```diff
@@ -23,20 +23,22 @@
...
+    <ClCompile Include="easylogging++.cpp">
+=======
+<<<<<<< HEAD
 <<<<<<< HEAD:3rdPartyLibs/easylogging++/samples/VC++/VCPP2015_Win32_Multithreaded/VCPP2015_Win32/VCPP2015_Win32.vcxproj.filters
     <ClCompile Include="easylogging++.cc">
 =======
     <ClCompile Include="easylogging++.cpp">
->>>>>>> Full project added:3rdPartyLibs/easylogging++/samples/VC++/VCPP2015_Win32/VCPP2015_Win32/VCPP2015_Win32.vcxproj.filters
+>>>>>>> 150391aee555e8ae6ccd79b7faefdbc6dd5c8ae4:3rdPartyLibs/easylogging++/samples/VC++/VCPP2015_Win32/VCPP2015_Win32/VCPP2015_Win32.vcxproj.filters
+=======
```
Comment (OleksiiiSemko) [inline 3rdPartyLibs/easylogging++/samples/VC++/VCPP2015_Win32/VCPP2015_Win32/VCPP2015_Win32.vcxproj.filters]: Let's add 3rdPartyLibs to gitignore

File: RealTimeFaceMaskDetector/TCPServer/SocketServer.cpp [hunk 6]
```diff
@@ -63,40 +63,69 @@ bool SocketServer::AcceptConnection()
...
+				std::thread th = std::thread([&]() {ReceiveMessage(ret_value);});
+				if(th.joinable())
+					th.detach();
+			}
+			else
+			{
+				ret_value =  false;
+				break;
+			}
+		}
```
Comment (OleksiiiSemko) [inline RealTimeFaceMaskDetector/TCPServer/SocketServer.cpp]: Can you create a separate function for this?

File: RealTimeFaceMaskDetector/TCPServer/SocketServer.cpp [hunk 7]
```diff
@@ -63,40 +63,69 @@ bool SocketServer::AcceptConnection()
...
 		// -- Insert photo --
-		Photo photo;
-		photo.path=R"(E:\Tolik\c++\Real-Time-Face-Mask-Detector\RealTimeFaceMaskDetector\TCPServer\)";
-		photo.name= "Avatar" ;
-		photo.extension="png";
+/*		Photo photo;
+		photo.path = R"(E:\Tolik\c++\Real-Time-Face-Mask-Detector\RealTimeFaceMaskDetector\TCPServer\)";
+		photo.name = "Avatar";
+		photo.extension = "png";
+*/
```
Comment (OleksiiiSemko) [inline RealTimeFaceMaskDetector/TCPServer/SocketServer.cpp]: Delete commented code

File: RealTimeFaceMaskDetector/TCPServer/SocketServer.cpp [hunk 8]
```diff
@@ -107,59 +136,158 @@ bool SocketServer::SendMessage()
...
 		{
-			recived_bytes_count += m_func_result;
-			m_buffer.insert(end(m_buffer), begin(temp_buffer), end(temp_buffer));
+			symbol = '_';
 		}
-		else if (m_func_result == 0)
+		else if(symbol == ':')
 		{
-			LOG_MSG << "Connection closing...";
+			symbol = '_';
```
Comment (OleksiiiSemko) [inline RealTimeFaceMaskDetector/TCPServer/SocketServer.cpp]: Can we move it to a separate function?

File: RealTimeFaceMaskDetector/TCPServer/SocketServer.cpp [hunk 9]
```diff
@@ -107,59 +136,158 @@ bool SocketServer::SendMessage()
...
+				//Writing in database
+				std::string data;
+				std::string cipher(m_buffer.begin(), m_buffer.end());
+				decryptor.Decrypt(cipher, data);
+				SendMessage();
+			}
+			else
+			{
+				client_connected = false;
+			}
```
Comment (OleksiiiSemko) [inline RealTimeFaceMaskDetector/TCPServer/SocketServer.cpp]: Can we move it to a separate function?

File: RealTimeFaceMaskDetector/TCPServer/WindowsService.cpp [hunk 10]
```diff
@@ -223,49 +229,54 @@ bool Service::Start()
...
-
+			if(!CreateServer())
+			{
+				LOG_WARNING << "Start: CreateServer: ERROR " << GetLastError();
+				is_started = false;
+			}
+			else
+			{
 			LOG_MSG << "Start: succeeded :)";
 		}
```
Comment (OleksiiiSemko) [inline RealTimeFaceMaskDetector/TCPServer/WindowsService.cpp]: Can we move it to a separate function?

**Review by OleksiiiSemko — CHANGES_REQUESTED** [review]

File: RealTimeFaceMaskDetector/TCPServer/SocketServer.cpp [hunk 11]
```diff
@@ -63,103 +61,169 @@ bool SocketServer::AcceptConnection()
...
-		else
+			if(ReceiveFullMessage())
+			{
+				SaveAndSendData();
+			}
+			else
+			{
+				client_connected = false;
+			}
+		}catch(const std::string& msg)
```
Comment (OleksiiiSemko) [inline RealTimeFaceMaskDetector/TCPServer/SocketServer.cpp]: Please move to a new line

File: RealTimeFaceMaskDetector/TCPServer/WindowsService.cpp [hunk 12]
```diff
@@ -223,49 +242,46 @@ bool Service::Start()
...
 		{
-			CreateServer();
-
-			LOG_MSG << "Start: succeeded :)";
+			TryCreateServer(is_started);
 		}
-
-		CloseServiceHandle(open_service);
+		CloseServiceHandle(handle_open_service);
 	}
```
Comment (OleksiiiSemko) [inline RealTimeFaceMaskDetector/TCPServer/WindowsService.cpp]: Please, move it to a separate method

File: 3rdPartyLibs/easylogging++/samples/VC++/VCPP2015_Win32/VCPP2015_Win32/VCPP2015_Win32.vcxproj.filters (hunk 5 above)
Comment (OleksiiiSemko) [reply 3rdPartyLibs/easylogging++/samples/VC++/VCPP2015_Win32/VCPP2015_Win32/VCPP2015_Win32.vcxproj.filters]: You should unstage 3PartyLibs firstly

File: RealTimeFaceMaskDetector/TCPServer/Main.cpp [hunk 13]
```diff
@@ -36,11 +36,7 @@ bool InteractWithServiceUsingCommandPrompt(const unsigned argc, const wchar_t* c
 {
 	if (argc > 1)
 	{
-		if (!wcscmp(argv[SERVICE_ARGUMENT], INSTALL.data()))
-		{
-			return Service::get_instance().Install();
-		}
-
+		if (!wcscmp(argv[SERVICE_ARGUMENT], INSTALL.data()))			return Service::get_instance().Install();
```
Comment (OleksiiiSemko) [inline RealTimeFaceMaskDetector/TCPServer/Main.cpp:39]: Please, move return to a new line and add curly braces

File: RealTimeFaceMaskDetector/TCPServer/SocketServer.cpp [hunk 14]
```diff
@@ -63,103 +61,169 @@ bool SocketServer::AcceptConnection()
...
-	std::vector<char> bytes_number;
-	bytes_number.resize(DEFAULT_BUFLEN);
-	recv(m_client_socket, &bytes_number[0], bytes_number.size(), 0);
 
-	return atoi(bytes_number.data());
+	if (AcceptConnection())
+	{
+		std::thread th = std::thread([&]() {ReceiveMessage(ret_value); });
+		if (th.joinable())
+			th.detach();
```
Comment (OleksiiiSemko) [inline RealTimeFaceMaskDetector/TCPServer/SocketServer.cpp]: I'm not sure, but it would be more readable if it was in a separate method

File: RealTimeFaceMaskDetector/TCPServer/SocketServer.cpp [hunk 15]
```diff
@@ -63,103 +61,169 @@ bool SocketServer::AcceptConnection()
...
-		}
-		else
+			if(ReceiveFullMessage())
+			{
+				SaveAndSendData();
+			}
+			else
+			{
+				client_connected = false;
+			}
```
Comment (OleksiiiSemko) [inline RealTimeFaceMaskDetector/TCPServer/SocketServer.cpp]: I know that's annoying) But a separate method add readability

File: RealTimeFaceMaskDetector/TCPServer/SocketServer.cpp [hunk 16]
```diff
@@ -63,103 +61,169 @@ bool SocketServer::AcceptConnection()
...
+}
+
+void SocketServer::SaveAndSendData()
+{
+	EncryptDecryptAES_ECBMode decryptor;
+	std::ofstream recv_data;
 
+	if (!OpenParticularFile(recv_data))
+	{
+		/*cannot open file error*/
```
Comment (OleksiiiSemko) [inline RealTimeFaceMaskDetector/TCPServer/SocketServer.cpp:194]: Why we need this if statement?

File: RealTimeFaceMaskDetector/TCPServer/SocketServer.cpp [hunk 17]
```diff
@@ -63,103 +61,169 @@ bool SocketServer::AcceptConnection()
...
+	try
+	{
+		sql_server->GetIniParams(CONFIG_FILE);
+
+		// -- Connect --
+		sql_server->Connect();
+		if (!sql_server->CheckTableExists("Photos"))
+		{
+			sql_server->CreatePhotosTable("Photos");
+		}
```
Comment (OleksiiiSemko) [inline RealTimeFaceMaskDetector/TCPServer/SocketServer.cpp]: Separate)

**Review by OleksiiiSemko — CHANGES_REQUESTED** [review]

File: RealTimeFaceMaskDetector/TCPServer/Main.cpp (hunk 13 above)
Comment (marko-yaciv) [reply RealTimeFaceMaskDetector/TCPServer/Main.cpp:39]: Here code looks bad but in the editor it's fine

File: RealTimeFaceMaskDetector/TCPServer/SocketServer.cpp (hunk 15 above)
Comment (marko-yaciv) [reply RealTimeFaceMaskDetector/TCPServer/SocketServer.cpp]: client_connected variable is used in that loop to hold connection with client , so that I didn't want to pass it anywhere else)

File: RealTimeFaceMaskDetector/TCPServer/SocketServer.cpp (hunk 16 above)
Comment (marko-yaciv) [reply RealTimeFaceMaskDetector/TCPServer/SocketServer.cpp:194]: Cause file can be not opened, so that photo will not be written there, and invalid path shouldn't be written to the database

File: 3rdPartyLibs/easylogging++/samples/VC++/VCPP2015_Win32/VCPP2015_Win32/VCPP2015_Win32.vcxproj.filters (hunk 5 above)
Comment (marko-yaciv) [reply 3rdPartyLibs/easylogging++/samples/VC++/VCPP2015_Win32/VCPP2015_Win32/VCPP2015_Win32.vcxproj.filters]: Do you mean using 'git rm 3rdPartyLibs'?

**Review by OleksiiiSemko — APPROVED** [review]
